File: core/src/probe.py
#!/usr/bin/env python3
"""ccwatch zero-install remote probe.

Piped over `ssh <host> python3 -` by ccwatchd, so nothing has to be installed
on the remote machine. Reads the remote's ~/.claude (or argv[1] /
$CLAUDE_CONFIG_DIR) and prints a Snapshot JSON matching ccwatch-core's serde
model exactly. Stdlib only; must stay compatible with python3.6+.
"""
import glob
import json
import os
import subprocess
import sys
import time
from datetime import datetime

ROOT = (
    sys.argv[1]
    if len(sys.argv) > 1
    else os.environ.get("CLAUDE_CONFIG_DIR", os.path.expanduser("~/.claude"))
)
NOW_MS = int(time.time() * 1000)
WINDOW_MS = 5 * 60 * 1000   # rate window, matches core default
IDLE_MS = 120 * 1000        # idle threshold, matches core default
MAX_TAIL = 16 * 1024 * 1024  # cap transcript read for huge histories
BUCKET_MS = 5 * 60 * 1000    # governor usage buckets
HORIZON_MS = 6 * 3600 * 1000  # how far back buckets are reported

# bucket_ts -> billable tokens, fed by every transcript scan
USAGE_BUCKETS = {}
# observed 429 timestamps (epoch ms) within the horizon
RATE_LIMITS = []
AGENT_TOOLS = ("Agent", "Task", "Workflow")


def zero_ledger():
    return dict(input=0, output=0, cache_write=0, cache_read=0,
                web_search=0, web_fetch=0, messages=0)


def alive(pid):
    try:
        os.kill(pid, 0)
        return True
    except Exception:
        return False


def parse_ts(ts):
    """ISO-8601 'Z' timestamp -> epoch ms, or None."""
    try:
        return int(
            datetime.fromisoformat(ts.replace("Z", "+00:00")).timestamp() * 1000
        )
    except Exception:
        return None


def parse_etime(e):
    """[[dd-]hh:]mm:ss -> seconds."""
    try:
        days = 0
        if "-" in e:
            d, e = e.split("-", 1)
            days = int(d)
        parts = [int(x) for x in e.split(":")]
        while len(parts) < 3:
            parts.insert(0, 0)
        return days * 86400 + parts[0] * 3600 + parts[1] * 60 + parts[2]
    except Exception:
        return 0


def process_table():
    """pid -> (ppid, cpu, rss_mb, run_secs, comm) for the whole system."""
    table = {}
    try:
        out = subprocess.run(
            ["ps", "-axo", "pid=,ppid=,pcpu=,rss=,etime=,comm="],
            capture_output=True, text=True, timeout=5,
        ).stdout
        for line in out.splitlines():
            f = line.split(None, 5)
            if len(f) < 6:
                continue
            table[int(f[0])] = (int(f[1]), float(f[2]), int(f[3]) // 1024,
                                parse_etime(f[4]), f[5])
    except Exception:
        pass
    return table


PROC_TABLE = process_table()


def children_of(root, cap=12):
    kids = {}
    for pid, (ppid, *_rest) in PROC_TABLE.items():
        kids.setdefault(ppid, []).append(pid)
    out, queue = [], [root]
    while queue:
        p = queue.pop()
        for c in kids.get(p, []):
            queue.append(c)
            ppid, cpu, rss, secs, comm = PROC_TABLE[c]
            out.append({
                "pid": c,
                "name": os.path.basename(comm),
                "cmd": comm,
                "cpu_pct": cpu,
                "rss_mb": rss,
                "run_secs": secs,
            })
    out.sort(key=lambda x: (-x["cpu_pct"], -x["rss_mb"]))
    return out[:cap]


def proc_stat(pid):
    """(cpu_pct, rss_mb) via ps — portable across macOS and Linux."""
    try:
        out = subprocess.run(
            ["ps", "-o", "pcpu=,rss=", "-p", str(pid)],
            capture_output=True, text=True, timeout=3,
        ).stdout.split()
        return float(out[0]), int(out[1]) // 1024
    except Exception:
        return 0.0, 0


# session id -> transcript path
transcripts = {}
for p in glob.glob(os.path.join(ROOT, "projects", "*", "*.jsonl")):
    transcripts[os.path.basename(p)[:-6]] = p


def scan_transcript(path):
    """Fold assistant usage; returns
    (ledger, window_billable, last_activity, model, agents)."""
    led = zero_ledger()
    window_billable = 0
    last_act = None
    model = None
    agents = {}  # tool_use id -> agent dict, insertion-ordered
    bg_ids = set()  # background launches: plain tool_results don't finish them
    pending = {}  # tool_use id -> (tool, detail, ts) awaiting a result
    try:
        size = os.path.getsize(path)
        fh = open(path, "rb")
        if size > MAX_TAIL:
            fh.seek(size - MAX_TAIL)
            fh.readline()  # drop the partial line
        for raw in fh:
            try:
                d = json.loads(raw)
            except Exception:
                continue
            ts = parse_ts(d.get("timestamp") or "")
            if d.get("apiErrorStatus") == 429 and ts and ts >= NOW_MS - HORIZON_MS:
                RATE_LIMITS.append(ts)
            m = d.get("message") or {}
            content = m.get("content")
            # Agent lifecycle: launches in assistant turns, completions via
            # tool_result carriers.
            if isinstance(content, list):
                for block in content:
                    if not isinstance(block, dict):
                        continue
                    if block.get("type") == "tool_use" and block.get("name") in AGENT_TOOLS:
                        inp = block.get("input") or {}
                        if inp.get("run_in_background") is True:
                            bg_ids.add(block.get("id") or "")
                        agents[block.get("id") or ""] = {
                            "id": block.get("id") or "",
                            "subagent_type": inp.get("subagent_type") or block.get("name"),
                            "description": inp.get("description") or "",
                            "model": inp.get("model"),
                            "state": "running",
                            "started_at": ts,
                            "tokens": zero_ledger(),
                            "tokens_per_min": 0.0,
                            "children": [],
                        }
                    elif block.get("type") == "tool_use" and block.get("name") not in AGENT_TOOLS:
                        inp = block.get("input") or {}
                        detail = ""
                        for key in ("file_path", "command", "pattern", "query", "url", "description"):
                            if isinstance(inp.get(key), str):
                                detail = inp[key][:100]
                                break
                        if ts:
                            pending[block.get("id") or ""] = (block.get("name") or "", detail, ts)
                    elif block.get("type") == "tool_result":
                        tid = block.get("tool_use_id")
                        pending.pop(tid, None)
                        # Background launches ack immediately; that's not done.
                        if tid in agents and tid not in bg_ids:
                            agents[tid]["state"] = "finished"
            tid = d.get("sourceToolUseID")
            pending.pop(tid, None)
            if tid in agents and tid not in bg_ids:
                agents[tid]["state"] = "finished"
            # Task-notifications are the authoritative background completion.
            if b"<task-notification>" in raw:
                text = raw.decode("utf-8", "replace")
                rest = text
                while "<tool-use-id>" in rest:
                    rest = rest.split("<tool-use-id>", 1)[1]
                    tid, _, rest = rest.partition("</tool-use-id>")
                    if tid in agents:
                        agents[tid]["state"] = "finished"

            if d.get("type") != "assistant":
                continue
            u = m.get("usage") or {}
            if not u:
                continue
            g = lambda k: u.get(k) or 0
            led["input"] += g("input_tokens")
            led["output"] += g("output_tokens")
            led["cache_write"] += g("cache_creation_input_tokens")
            led["cache_read"] += g("cache_read_input_tokens")
            srv = u.get("server_tool_use") or {}
            led["web_search"] += srv.get("web_search_requests") or 0
            led["web_fetch"] += srv.get("web_fetch_requests") or 0
            led["messages"] += 1
            # Skip internal markers like "<synthetic>".
            if m.get("model") and not m["model"].startswith("<"):
                model = m["model"]
            if ts:
                last_act = max(last_act or 0, ts)
                billable = (
                    g("input_tokens")
                    + g("output_tokens")
                    + g("cache_creation_input_tokens")
                )
                if ts >= NOW_MS - WINDOW_MS:
                    window_billable += billable
                if ts >= NOW_MS - HORIZON_MS:
                    bucket = ts - (ts % BUCKET_MS)
                    USAGE_BUCKETS[bucket] = USAGE_BUCKETS.get(bucket, 0) + billable
        fh.close()
    except Exception:
        pass
    activity = [
        {"tool": t, "detail": det, "since_ms": ts}
        for t, det, ts in sorted(pending.values(), key=lambda x: x[2])
        if NOW_MS - ts < 30 * 60 * 1000
    ][:6]
    return led, window_billable, last_act, model, list(agents.values()), activity


def read_tasks(sid):
    out = []
    files = glob.glob(os.path.join(ROOT, "tasks", sid, "*.json"))

    def order(p):
        stem = os.path.basename(p)[:-5]
        return int(stem) if stem.isdigit() else 1 << 60

    for tf in sorted(files, key=order):
        try:
            td = json.load(open(tf))
        except Exception:
            continue
        if not td.get("subject"):
            continue
        out.append({
            "subject": td["subject"],
            "status": td.get("status", ""),
            "blocked": bool(td.get("blockedBy")),
            "active_form": td.get("activeForm"),
        })
    return out


sessions = []
for f in sorted(glob.glob(os.path.join(ROOT, "sessions", "*.json"))):
    try:
        meta = json.load(open(f))
    except Exception:
        continue
    sid = meta.get("sessionId")
    pid = meta.get("pid")
    if not sid or not pid or not alive(pid):
        continue

    led, window_billable, last_act, model, agents, activity = (
        scan_transcript(transcripts[sid])
        if sid in transcripts
        else (zero_ledger(), 0, None, None, [], [])
    )
    tpm = window_billable / (WINDOW_MS / 60000.0)
    last = last_act or meta.get("startedAt")
    state = "running" if (last is None or NOW_MS - last <= IDLE_MS) else "idle"
    cpu, rss = proc_stat(pid)

    sessions.append({
        "id": sid,
        "name": meta.get("name") or sid,
        "cwd": meta.get("cwd", ""),
        "pid": pid,
        "kind": meta.get("kind") or "interactive",
        "entrypoint": meta.get("entrypoint", ""),
        "version": meta.get("version", ""),
        "model": model,
        "state": state,
        "started_at": meta.get("startedAt"),
        "last_activity": last,
        "tokens": led,
        "tokens_per_min": tpm,
        "cpu_pct": cpu,
        "rss_mb": rss,
        "agents": agents,
        "tasks": read_tasks(sid),
        "watchers": [],
        "activity": activity,
        "processes": children_of(pid),
        "host": {"kind": "local"},
        "remote_name": None,
    })

agg = dict(input=0, cache_write=0, cache_read=0)
total_tokens = 0
for s in sessions:
    t = s["tokens"]
    total_tokens += t["input"] + t["output"] + t["cache_write"] + t["cache_read"]
    for k in agg:
        agg[k] += t[k]
denom = agg["input"] + agg["cache_write"] + agg["cache_read"]
print(json.dumps({
    "generated_at": NOW_MS,
    "sessions": sessions,
    "alerts": [],
    "usage_buckets": sorted(USAGE_BUCKETS.items()),
    "rate_limits": sorted(set(RATE_LIMITS)),
    "totals": {
        "active_sessions": len(sessions),
        "tokens_per_min": sum(s["tokens_per_min"] for s in sessions),
        "total_tokens": total_tokens,
        "cache_hit_pct": (agg["cache_read"] / denom * 100.0) if denom else 0.0,
    },
}))

File: core/tests/probe.rs
//! The zero-install SSH probe must emit JSON that parses as our exact
//! [`Snapshot`] schema. Runs the real `probe.py` under python3 against a
//! fixture tree — the same way the daemon runs it over ssh.

use ccwatch_core::model::{Host, SessionState, Snapshot};
use ccwatch_core::remote::{CommandRunner, SystemRunner, PROBE_PY};
use chrono::TimeZone;
use std::time::Duration;

fn rfc3339(ms: i64) -> String {
    chrono::Utc
        .timestamp_millis_opt(ms)
        .unwrap()
        .to_rfc3339_opts(chrono::SecondsFormat::Millis, true)
}

#[test]
fn probe_emits_valid_snapshot() {
    let root = std::env::temp_dir().join(format!("ccw-probe-{}", std::process::id()));
    let _ = std::fs::remove_dir_all(&root);

    let pid = std::process::id() as i32; // alive on "the remote"
    let sid = "probe-sess";
    let now_ms = chrono::Utc::now().timestamp_millis();

    std::fs::create_dir_all(root.join("sessions")).unwrap();
    std::fs::write(
        root.join("sessions").join(format!("{pid}.json")),
        format!(
            r#"{{"pid":{pid},"sessionId":"{sid}","cwd":"/remote/proj","kind":"interactive","entrypoint":"cli","version":"2.1.0","name":"probed","startedAt":{}}}"#,
            now_ms - 60_000
        ),
    )
    .unwrap();

    let proj = root.join("projects").join("-remote-proj");
    std::fs::create_dir_all(&proj).unwrap();
    let mut lines = String::new();
    // One old message (outside the 5-min window) and two fresh ones.
    for (age_ms, out) in [(600_000i64, 7_000u64), (30_000, 1_000), (5_000, 2_000)] {
        let v = serde_json::json!({
            "type": "assistant",
            "timestamp": rfc3339(now_ms - age_ms),
            "message": {
                "model": "claude-opus-4-8",
                "usage": {
                    "input_tokens": 100,
                    "output_tokens": out,
                    "cache_creation_input_tokens": 50,
                    "cache_read_input_tokens": 9_000,
                    "server_tool_use": {"web_search_requests": 1, "web_fetch_requests": 0}
                }
            }
        });
        lines.push_str(&format!("{v}\n"));
    }
    // An agent launch (running: no tool_result follows) and a 429 event.
    let agent_line = serde_json::json!({
        "type": "assistant",
        "timestamp": rfc3339(now_ms - 20_000),
        "message": {
            "model": "claude-opus-4-8",
            "content": [{"type": "tool_use", "id": "toolu_r1", "name": "Agent",
                         "input": {"description": "remote scan", "subagent_type": "Explore"}}]
        }
    });
    lines.push_str(&format!("{agent_line}\n"));
    let rl_line = serde_json::json!({
        "type": "assistant",
        "timestamp": rfc3339(now_ms - 10_000),
        "apiErrorStatus": 429,
        "isApiErrorMessage": true,
        "message": {"content": "rate limited"}
    });
    lines.push_str(&format!("{rl_line}\n"));
    // Background agent: immediate ack must NOT finish it; a task-notification
    // for a second one does.
    for (id, desc) in [("toolu_bg1", "bg still running"), ("toolu_bg2", "bg completed")] {
        let launch = serde_json::json!({
            "type": "assistant",
            "timestamp": rfc3339(now_ms - 15_000),
            "message": {"content": [{"type": "tool_use", "id": id, "name": "Agent",
                "input": {"description": desc, "subagent_type": "claude", "run_in_background": true}}]}
        });
        let ack = serde_json::json!({
            "type": "user",
            "message": {"content": [{"type": "tool_result", "tool_use_id": id, "content": "started"}]}
        });
        lines.push_str(&format!("{launch}\n{ack}\n"));
    }
    lines.push_str(&format!(
        "{}\n",
        serde_json::json!({
            "type": "user",
            "timestamp": rfc3339(now_ms - 5_000),
            "message": {"content": "<task-notification><task-id>t1</task-id><tool-use-id>toolu_bg2</tool-use-id><status>completed</status></task-notification>"}
        })
    ));
    // An in-flight Edit: tool_use with no tool_result.
    lines.push_str(&format!(
        "{}\n",
        serde_json::json!({
            "type": "assistant",
            "timestamp": rfc3339(now_ms - 3_000),
            "message": {"content": [{"type": "tool_use", "id": "toolu_edit", "name": "Edit",
                "input": {"file_path": "/remote/x.rs", "old_string": "a", "new_string": "b"}}]}
        })
    ));
    std::fs::write(proj.join(format!("{sid}.jsonl")), lines).unwrap();

    let tasks = root.join("tasks").join(sid);
    std::fs::create_dir_all(&tasks).unwrap();
    std::fs::write(
        tasks.join("1.json"),
        r#"{"subject":"remote todo","status":"in_progress","blockedBy":["0"]}"#,
    )
    .unwrap();

    // Run exactly as the daemon would over ssh: `python3 - <root>`, probe on stdin.
    let argv: Vec<String> = vec![
        "python3".into(),
        "-".into(),
        root.to_str().unwrap().into(),
    ];
    let out = SystemRunner
        .run(&argv, Some(PROBE_PY), Duration::from_secs(15))
        .expect("probe should run");

    let snap: Snapshot = serde_json::from_str(out.trim()).expect("probe output must parse");

    assert_eq!(snap.sessions.len(), 1);
    let s = &snap.sessions[0];
    assert_eq!(s.id, sid);
    assert_eq!(s.name, "probed");
    assert_eq!(s.pid, Some(pid));
    assert_eq!(s.state, SessionState::Running);
    assert_eq!(s.model.as_deref(), Some("claude-opus-4-8"));
    assert!(matches!(s.host, Host::Local)); // retagged by fetch_remote later

    // Full-history ledger.
    assert_eq!(s.tokens.messages, 3);
    assert_eq!(s.tokens.output, 10_000);
    assert_eq!(s.tokens.input, 300);
    assert_eq!(s.tokens.cache_write, 150);
    assert_eq!(s.tokens.cache_read, 27_000);
    assert_eq!(s.tokens.web_search, 3);

    // Rate counts only the two in-window messages:
    // (100+1000+50) + (100+2000+50) = 3300 billable over 5 min = 660/min.
    assert!(
        (s.tokens_per_min - 660.0).abs() < 1.0,
        "expected ~660 tok/min, got {}",
        s.tokens_per_min
    );

    // Agents: foreground running + two background with correct lifecycle.
    use ccwatch_core::model::AgentState;
    assert_eq!(s.agents.len(), 3);
    let by_desc = |d: &str| s.agents.iter().find(|a| a.description == d).unwrap();
    assert!(matches!(by_desc("remote scan").state, AgentState::Running));
    assert!(
        matches!(by_desc("bg still running").state, AgentState::Running),
        "background ack must not finish the agent"
    );
    assert!(
        matches!(by_desc("bg completed").state, AgentState::Finished),
        "task-notification must finish the background agent"
    );

    // The 429 came through.
    assert_eq!(snap.rate_limits.len(), 1);

    // In-flight tool call (Edit with no result yet) → live activity.
    assert!(
        s.activity.iter().any(|a| a.tool == "Edit" && a.detail == "/remote/x.rs"),
        "in-flight Edit should be activity: {:?}",
        s.activity
    );

    // Child processes came through (the probe itself is a child of our pid).
    assert!(
        s.processes.iter().any(|p| p.name.to_lowercase().contains("python")),
        "probe should see itself in the session's process tree: {:?}",
        s.processes
    );

    // Tasks came through.
    assert_eq!(s.tasks.len(), 1);
    assert_eq!(s.tasks[0].subject, "remote todo");
    assert!(s.tasks[0].blocked);

    // Totals coherent.
    assert_eq!(snap.totals.active_sessions, 1);
    assert!(snap.totals.cache_hit_pct > 90.0);

    // Governor usage buckets: all three messages are within the 6h horizon;
    // billable per message = 100 + out + 50.
    let bucket_sum: u64 = snap.usage_buckets.iter().map(|(_, v)| v).sum();
    assert_eq!(bucket_sum, 300 + 10_000 + 150, "buckets carry billable tokens");
    assert!(snap.usage_buckets.windows(2).all(|w| w[0].0 < w[1].0), "buckets sorted");

    let _ = std::fs::remove_dir_all(&root);
}
